fix(docs): add docs for openvino (#2051)
lf-edge/ekuiper: docs/en_US/guide/ai/python_openvino_tutorial.md

# Running AI Algorithms with Python Function Plugins

[LF Edge eKuiper](https://www.lfedge.org/projects/ekuiper/) is an edge lightweight IoT data analytics / streaming
software which can be run at all kinds of resource constrained IoT devices.

[OpenVINO](https://www.intel.com/content/www/us/en/developer/tools/openvino-toolkit/overview.html) is an open source toolkit that makes it easier to write once, deploy anywhere.

> It converts and optimizes models trained using popular frameworks like TensorFlow*, PyTorch*, and Caffe*. Deploy across a mix of Intel hardware and environments, on-premise and on-device, in the browser, or in the cloud.

Reviewer: Why add * after TensorFlow, PyTorch etc..

    ```python
    if __name__ == '__main__':
        # Define the plugin
        c = PluginConfig("defect", {}, {},
                         {"inference": lambda: inferenceIns})
        # Start the plugin instance
        plugin.start(c)
    ```

3. Create a plugin description file in JSON format to define the metadata of the plugin. The file name must be the same as the plugin name, i.e. `defect.json`. The function names defined therein must correspond exactly to the entry file, and the contents of the file are as follows. Where, executable is used to define the name of the plugin's executable entry file.

    ```json
    {
      "version": "v1.0.0",
      "language": "python",
      "executable": "main.py",
      "sources": [
      ],
      "sinks": [
      ],
      "functions": [
        "inference"
      ]
    }
    ```

At this point we have completed the development of the plugin, next we just need to package all the files in the directory into a zip file. The zip file should have a file structure similar to:

- inference.py
- requirements.txt
- install.sh
- main.py
- defect.json
- models

>   - model.bin

Reviewer: Are these models public? Can we disclose them? Can we just put a link here instead of ship them?

Author: Intel do not have  a public link that holds these models

> # Running AI Algorithms with Python Function Plugins

Reviewer 2: ```suggestion
# Running OpenVINO Algorithms with Python Function Plugin
```

[LF Edge eKuiper](https://www.lfedge.org/projects/ekuiper/) is an edge lightweight IoT data analytics / streaming

> software which can be run at all kinds of resource constrained IoT devices.

Reviewer 2: ```suggestion
software which can be run at all kinds of resource-constrained IoT devices.
```

> [OpenVINO](https://www.intel.com/content/www/us/en/developer/tools/openvino-toolkit/overview.html) is an open source toolkit that makes it easier to write once, deploy anywhere.

Reviewer 2: ```suggestion
[OpenVINO](https://www.intel.com/content/www/us/en/developer/tools/openvino-toolkit/overview.html) is an open source toolkit that provides a "write once, deploy anywhere" functionality.
```

> It converts and optimizes models trained using popular frameworks like TensorFlow, PyTorch, and Caffe. Deploy across a mix of Intel hardware and environments, on-premise and on-device, in the browser, or in the cloud.

Reviewer 2: ```suggestion
It converts and optimizes models trained using popular frameworks like TensorFlow, PyTorch, and Caffe, and offers flexibility in deployment, supporting a mix of Intel hardware and environments, on-premise, on-device, web browser, and cloud environment.
```

> The example code and models for OpenVINO take reference from this [post](https://www.intel.cn/content/www/cn/zh/developer/articles/reference-implementation/industrial-surface-defect-detection.html), users can get more detail information from that link.

Reviewer 2: ```suggestion
The example code and models for OpenVINO take reference from [Intel - Industrial Surface Defect Detection Reference Implementation](https://www.intel.com/content/www/us/en/developer/articles/reference-implementation/industrial-surface-defect-detection.html).
```

> By integrating eKuiper and OpenVINO, users can analyze the data more easily.

Reviewer 2: By combining eKuiper and OpenVINO, data analysis can become more accessible and efficient.

> In this tutorial, we will walk you through building an AI-enabled approach to segment defects from the surface powered by eKuiper and OpenVINO.

Reviewer 2: ```suggestion
This tutorial will guide you through creating an AI-based system for defect detection on surfaces, utilizing the power of eKuiper and OpenVINO. 
```

> The completed plugin package can be downloaded [here](https://github.com/lf-edge/ekuiper/blob/master/docs/resources/openvinoProject.zip), which also contains the full source code.

Reviewer 2: ```suggestion
A complete plugin package, including full source code, can be downloaded from [eKuiper Resources page](https://github.com/lf-edge/ekuiper/blob/master/docs/resources/openvinoProject.zip).
```

## Prerequisite

Before starting the tutorial, please prepare the following products or environments.

> 1. Install the Python 3.x environment.

Reviewer 2: ```suggestion
1. Python 3.x
```

> 1. Install the Python 3.x environment.

Reviewer 2: ```suggestion
1. Install the Python 3.x environment. And set the `pythonBin` under the portable configuration of [eKuiper configuration file](../../configuration/global_configurations.md#portable-plugin-configurations) to match your Python command (e.g., 'python3' if applicable).
```

> 2. Install the opencv-python, numpy and openvino packages via `pip install opencv-python==4.7.0.* openvino==2023.0.0 numpy==1.24.3`.

Reviewer 2: ```suggestion
2. Install the opencv-python, numpy and openvino packages via the following pip command

```python
pip install opencv-python==4.7.0.* openvino==2023.0.0 numpy==1.24.3
```
```

> By default, the portable plugin for eKuiper will run with the `python` command. If your environment does not support the `python` command, please use the [configuration file](../../configuration/global_configurations.md#portable-plugin-configurations) to modify the Python command, such as `python3`.

Reviewer 2: ```suggestion

```

Reviewer 2: this one is still about the python environment, that's why I combined this one with Prerequisite 1

> If you are developing with Docker, you can use the `lfedge/ekuiper:<tag>-slim-python` version. This version includes both the eKuiper and python environments.

Reviewer 2: For Docker users:

Use the lfedge/ekuiper:<tag>-slim-python Docker image which includes eKuiper and Python environments. If this Docker image lacks the required Python packages, install them using the pip command mentioned earlier, either by extending the Dockerfile or running the command inside the Docker container.

> If you are developing with Docker, you can use the `lfedge/ekuiper:<tag>-slim-python` version. This version includes both the eKuiper and python environments.

Reviewer 2: ```suggestion
For Docker users:

Use the lfedge/ekuiper:<tag>-slim-python Docker image which includes eKuiper and Python environments. If this Docker image lacks the required Python packages, install them using the pip command mentioned earlier, either by extending the Dockerfile or running the command inside the Docker container.
```

## Develop the Plugin

> To integrate eKuiper with OpenVINO, we will develop a custom eKuiper function plugin to use with eKuiper rules. We will create the `inference` function whose input is base64 encoded image data and whose output is json dictionary data containing the number of segment defects 、the processed image data and the cost for the inference.

Reviewer 2: ```suggestion
To integrate eKuiper with OpenVINO, we will develop a custom eKuiper function plugin to use with eKuiper rules. This plugin will include an inference function that takes base64 encoded image data as input and outputs a JSON dictionary. The output dictionary will contain information such as the number of segmented defects, processed image data, and the cost of the inference process.
```

To develop the function plugin, we need to：

> 1. implement the business logic in Python and wrap it as an eKuiper function.

Reviewer 2: ```suggestion
1. Implement the business logic in Python and wrap it as an eKuiper function.
```

2. Package the relevant files according to the plugin format.

> Create Python files that implement the extended interface (source, sink, or function). In this tutorial, we are developing a function plugin, so we need to implement the function extension interface.

Reviewer 2: ```suggestion
3. Create a Python file that implements the extended interface (source, sink, or function):
```

> - Writing Python segment defects functions

Reviewer 2: for these 2, let's make them substeps of step 3.

> - Wrapping an existing function as an eKuiper function plugin

Reviewer 2: same as above

> ### Implement the business logic

Reviewer 2: ```suggestion
### Implement the Business Logic
```

> Our target function needs to take the base64 encoded image data as an input parameter, perform image preprocessing, load the OpenVINO models, call the OpenVINO for inference, extract the inference result, and output it. We need to implement this function using Python in the same way as we would write a normal Python function.

Reviewer 2: ```suggestion
Our goal is to create a function that accepts base64 encoded image data, conducts image preprocessing, loads OpenVINO models, makes inference calls through OpenVINO, extracts inference results, and provides an output. This function will be implemented as a standard Python function.
```

    # Start sync inference

    t0 = time()
    img_str = base64.b64decode(file_bytes.encode("ascii"))
    ndarray = np.fromstring(img_str, np.uint8)
    frame = cv2.imdecode(ndarray, cv2.IMREAD_COLOR)  #BGR
    frame = cv2.resize(frame, (w, h))
    org_img = frame.copy()
    frame = frame.transpose((2, 0, 1))
    images = np.expand_dims(frame, axis=0)
    pred = exec_net.infer(inputs={input_blob: images})
    infer_time = ((time() - t0) * 1000)
    result = np.squeeze(pred[output_blob])
    thresh = 0.5
    result[result < thresh] = 0
    result[result > thresh] = 255
    result = result.astype(np.uint8)
    contours, hierarchy = cv2.findContours(result, cv2.RETR_TREE,
                                           cv2.CHAIN_APPROX_SIMPLE)

    pred_mask = np.zeros((h, w, 3))
    pred_mask[result < 0.5] = (0, 0, 0)
    pred_mask[result >= 0.5] = (255, 255, 255)

    base64_str = cv2.imencode('.jpg', np.hstack((org_img, pred_mask)))[1].tostring()
    base64_byte = base64.b64encode(base64_str)
    b64str = base64_byte.decode()

    result = {
        "inference time": infer_time,
        "defect": len(contours),
        "base64": b64str
    }

    return result
```

> The above code is only related to the business logic and can be tested without calling eKuiper's SDK. We just need to make sure that the input and output is of a type that can be converted to JSON format. For example, if the return value is a numpy array, it needs to be converted to a list type first. Developers can add main functions or unit tests to their business logic file or to another file for testing. For example, the following main function can be used to test the above business logic.

Reviewer 2: ```suggestion
The above code is only related to the business logic and can be tested without calling eKuiper's SDK. We just need to make sure that the input and output is of a type that can be converted to JSON format. For example, if the return value is a numpy array, it needs to be converted to a list type first. Developers can add main functions or unit tests to their business logic file or to another file for testing. For example, the following main function can be used to test the above business logic.
```

> make sure the model.xml and model.bin file are in the models directory.

Reviewer 2: ```suggestion
Ensure the `model.xml` and `model.bin` files are in the `models` directory.
```

```python
# To test the logic
if __name__ == '__main__':
    file_name = "test.jpg"
    with open(file_name, "rb") as f:
        cwd = './'
        result = inference(base64.b64encode(f.read()))
        print(json.dumps(result))
```

This file uses the `test.jpg` image file as input, calls the inference function to test it, and converts the result to a json string and prints it. This allows us to see if the function works as expected.

> ### Plugin Implementation

Reviewer 2: ```suggestion
### Build the Plugin
```

> Like native plugins, Python plugins need to implement the corresponding interfaces; Python plugins also support the Source, Sink and Function interfaces, [interface definition](../../extension/portable/python_sdk.md#development) is similar to the native plugins. Here, what we need to implement is the function interface.

Reviewer 2: ```suggestion
Similar to native plugins, Python plugins need to implement the corresponding interfaces, including the Source, Sink, and Function interfaces. The [Interface definition](../../extension/portable/python_sdk.md#development) for Python plugins is similar to the native plugins.
```

> Create the `inference.py` function to wrap the functions implemented in the previous section. Import the Function class from eKuiper's plugin SDK and create the corresponding implementation class. The validate function is used to validate the parameters; is_aggregate is used to define whether the function is an aggregate function. The key implementations are in the exec function. Here, we take the data in the eKuiper stream as an argument, call the logic implemented above, and return the result to eKuiper.

Reviewer 2: ```suggestion
Create the `inference.py` function to encapsulate the previously implemented functionalities. Import the Function class from eKuiper's plugin SDK and create the corresponding implementation class. The `validate` function is used to validate the parameters; `is_aggregate` is to define whether the function is an aggregate function. The core implementations are in the `exec` function. Here, we take the data in the eKuiper stream as an argument, call the previously implemented logic, and return the result to eKuiper.
```

```python
from typing import List, Any

from ekuiper import Function, Context

class InferenceFunc(Function):

    def __init__(self):
        pass

    def validate(self, args: List[Any]):
        if len(args) != 1:
            return "invalid arg length"
        return ""

    def exec(self, args: List[Any], ctx: Context):
        logging.debug("executing inference")
        return inference(args[0])

    def is_aggregate(self):
        return False

inferenceIns = InferenceFunc()
```

> Once the code is implemented, we also need to add a description file for each function, which is placed in the functions directory, in this case we create the `defect.json` file. This file will help the eKuiper manager to automatically produce the UI associated with the plugin.

Reviewer 2: ```suggestion
Once the code is implemented, we also need to add a description file for each function, which is placed in the functions directory, in this case, we create the `defect.json` file. 
```

### Plugin Packaging

> At this point, we have completed the development of the main functionality, and next we need to package these files into a plugin format. Plugin packaging requires several steps to be completed.

Reviewer 2: ```suggestion
At this point, we have completed the development of the main functionality, and next we need to package these files into a plugin format:
```

> 1. If the plugin has additional dependencies, you need to create the dependency installation script `install.sh`. When the plugin is installed, eKuiper will look for an installation script file `install.sh` in the plugin package and execute it if there is one. In this case, we create a `requirements.txt` file listing all the dependency packages. The installation of the dependencies is done in `install.sh` by calling `pip install -r $cur/requirements.txt`. For other plugins, you can reuse this script to update `requirements.txt` if you have no special requirements.

Reviewer 2: ```suggestion
1. **Managing Plugin Dependencies**: If the plugin has additional dependencies, you need to create the dependency installation script `install.sh`. When the plugin is installed, eKuiper will search for this script within the plugin package and executes it if found. In this case, we create a `requirements.txt` file listing all the dependency packages. The installation of the dependencies is done in `install.sh` by calling `pip install -r $cur/requirements.txt`. For other plugins, you can reuse this script to update `requirements.txt` if you have no special requirements.
```

> 2. Create a Python entry file that exposes all the implemented interfaces. Because multiple extensions can be implemented in a single plugin, you need an entry file that defines the implementation classes for each extension. The content is a main function, which is the entry point for the plugin runtime. It calls the methods in the SDK to define the plugin, including the plugin name, and a list of keys for the source, sink, and function implemented in the plugin. Here only a function plugin named `inference` is implemented, and its corresponding implementation method is `inferenceIns`. The Python plug-in process is independent of the eKuiper main process.

Reviewer 2: ```suggestion
2. **Create a Python Entry File**: Because multiple extensions can be implemented in a single plugin, you need an entry file that defines the implementation classes for each extension. The entry file is a main function, which is the entry point for the plugin runtime. It calls the methods in the SDK to define the plugin, including the plugin name, and a list of keys for the implemented source, sink, and function. Here only a function plugin named `inference` is implemented, with `inferenceIns` as its corresponding implementation method. The Python plug-in process operates independently from the eKuiper main process.
```

    ```python
    if __name__ == '__main__':
        # Define the plugin
        c = PluginConfig("defect", {}, {},
                         {"inference": lambda: inferenceIns})
        # Start the plugin instance
        plugin.start(c)
    ```

> 3. Create a plugin description file in JSON format to define the metadata of the plugin. The file name must be the same as the plugin name, i.e. `defect.json`. The function names defined therein must correspond exactly to the entry file, and the contents of the file are as follows. Where, executable is used to define the name of the plugin's executable entry file.

Reviewer 2: ```suggestion
3. **Establish a Plugin Description File**: Create a plugin description file in JSON format to define the metadata of the plugin. The file name must match the plugin name, i.e. `defect.json`. The function names defined within the file must align precisely with those in the entry file. The `executable` field is used to define the name of the plugin's executable entry file.
```

    ```json
    {
      "version": "v1.0.0",
      "language": "python",
      "executable": "main.py",
      "sources": [
      ],
      "sinks": [
      ],
      "functions": [
        "inference"
      ]
    }
    ```

At this point we have completed the development of the plugin, next we just need to package all the files in the directory into a zip file. The zip file should have a file structure similar to:

- inference.py
- requirements.txt
- install.sh
- main.py
- defect.json
- models
  - model.bin
  - model.xml
- functions
  - defect.json

> ## Plugin Installation

Reviewer 2: ```suggestion
## Install the Plugin
```

> As the same with installing native plugins, we can also install Python plugins via the eKuiper manager UI or the REST API. To use the REST API, upload the zip file packaged above to the machine where eKuiper is located. Then use the following API to install it:

Reviewer 2: ```suggestion
he same as installing native plugins, the Python plugins can be installed with REST API. To use the REST API, upload the zip file packaged above to the machine where eKuiper is located. Then use the following API to install it:
```

```text
### Install pyai plugin
POST http://{{host}}/plugins/functions
Content-Type: application/json

{"name":"defect", "file": "file:///tmp/defect.zip"}
```

The installation process requires an Internet connection to download dependencies, including `ffmpeg libsm6 libxext6`, which may take a long time depending on the network conditions.

> ## Run the plugin

Reviewer 2: ```suggestion
## Start the Plugin
```

> Once the plugin installed, we can use it in our rule. We will create a rule to receive image byte data from a mqtt topic and do the inference for the image by OpenVINO model.

Reviewer 2: ```suggestion
After installing the plugin, it can be incorporated into our rule. We'll create a rule that receives image byte data from an MQTT topic and performs inference on the image using the OpenVINO model.
```

> ### Define the stream

Reviewer 2: ```suggestion
### Define the Stream
```

Define the stream by eKuiper rest API. We create a stream named openvino_demo and the topic is openvino_demo.

```shell
POST http://{{host}}/streams
Content-Type: application/json

{"sql":"CREATE STREAM openvino_demo () WITH (DATASOURCE=\"openvino_demo\")"}
```

> ### Define the rule

Reviewer 2: ```suggestion
### Define the Rule
```

> Define the rule by eKuiper rest API.  We will create a rule named ruleOp. We just read the base64 encoded images from openvino_demo stream and run the custom function *inference* against it.  

Reviewer 2: ```suggestion
Define the rule by eKuiper rest API. This rule will read base64 encoded images from the `openvino_demo` stream and apply the custom function `inference` to it.
```

> It will send out the base64 encoded origin and processed image to topic *ekuiper/defect* when the segment defects numbers not be zero.

Reviewer 2: ```suggestion
If the number of segmented defects is non-zero, it will dispatch the base64 encoded original and processed images to the `ekuiper/defect` topic.
```

Reviewer 2: I think these 2 sentences belong to one paragraph.

> sends it to the MQTT topic `ekuiper/labels`.

Reviewer 2: ```suggestion

```

```shell
POST http://{{host}}/rules
Content-Type: application/json

{
  "id": "ruleOp",
  "sql": "SELECT image as origin, inference(image)->base64 as afterProcess from openvino_demo WHERE inference(image)->defect >=0",
  "actions": [
   {
      "mqtt":{
        "server": "tcp://emqx.io:1883",
        "sendSingle": true,
        "topic": "ekuiper/defect"
      }
    }
  ]
}
```

> ### Feed the data

Reviewer 2: ```suggestion
### Feed the Data
```

> Here we create a python program to send image data to the openvino_demo topic to be processed by the rule.

Reviewer 2: ```suggestion
Here we create a Python program to send image data to the `openvino_demo` topic and to be processed by the rule.
```

Users can get the full code [here](https://github.com/lf-edge/ekuiper/blob/master/docs/resources/openvinoProject.zip)

```python
def publish(client):
    topic = "openvino_demo"
    msg_count = 1
    while True:
        time.sleep(1)
        encoded = base64.b64encode(open('./1.png', 'rb').read()).decode()
        # open('1.txt', 'wb').write(encoded)  # 保存

        request_data = {
            "image": encoded,
        }

        payload = json.dumps(request_data)
        result = client.publish(topic, payload)
        # result: [0, 1]
        status = result[0]
        if status == 0:
            print(f"Send payload to topic `{topic}`")
        else:
            print(f"Failed to send message to topic {topic}")
        msg_count += 1
        if msg_count > 5:
            break
```

> ### Check the result

Reviewer 2: ```suggestion
### Check the Result
```

> Users can subscribe to the *ekuiper/defect* to get the notification when the images have segment defect.

Reviewer 2: ```suggestion
Users can subscribe to the `ekuiper/defect` topic to get a notification when the images exhibit segmented defects.
```

## Conclusion

> In this tutorial, we guide you through building custom eKuiper Python plugins that do segment defects for real-time image streams using OpenVINO models. Developers can replace the models with their own desired models to implement their own plug-ins.

Reviewer 2: ```suggestion
In this tutorial, we guide you through building custom eKuiper Python plugins that do defect segmentation for real-time image streams using OpenVINO models. Developers can replace the models with their own desired models to develop customized plugins.
```
